### PR TITLE
No global eval settings in `libnixexpr`

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Progress on #5638

# Context

There is still a global eval settings, but it pushed down into `libnixcmd`, which is a lot less bad a place for this sort of thing.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr-c/nix_api_expr.cc [hunk 1]
```diff
@@ -106,7 +107,15 @@ EvalState * nix_state_create(nix_c_context * context, const char ** lookupPath_c
...
-        return new EvalState{nix::EvalState(nix::LookupPath::parse(lookupPath), store->ptr)};
+        // FIXME leak
+        auto * evalSettings = new nix::EvalSettings{nix::settings.readOnlyMode};
+
+        return new EvalState{
+            nix::EvalState(
+                nix::LookupPath::parse(lookupPath),
+                store->ptr,
+                *evalSettings)
+        };
```
Comment (Ericson2314) [inline src/libexpr-c/nix_api_expr.cc:118]: @roberth I don't think this is very good, but I am not sure what we should do here instead?

I do like the forward compat of the C bindings not exposing any settings or reading the global ones.

File: src/libexpr-c/nix_api_expr.cc [hunk 2]
```diff
@@ -106,7 +107,20 @@ EvalState * nix_state_create(nix_c_context * context, const char ** lookupPath_c
...
+        return new (p) EvalState {
+            .readOnlyMode = false,
+            .settings = nix::EvalSettings{
+                p2->readOnlyMode,
+            },
+            .state = nix::EvalState(
+                nix::LookupPath::parse(lookupPath),
+                store->ptr,
+                p2->settings),
+        };
```
Comment (Ericson2314) [inline src/libexpr-c/nix_api_expr.cc:123]: This version is better than what I originally did, but still should double check with @roberth.

**Comment by edolstra** [comment]
Is there a use case where we could have multiple `EvalState` instances with different settings?

**Comment by Ericson2314** [comment]
@edolstra Evaluating completely independent things. Could be part of an application, or it could be part of the unit tests.

File: src/libexpr-c/nix_api_expr.cc [hunk 3]
```diff
@@ -106,7 +107,21 @@ EvalState * nix_state_create(nix_c_context * context, const char ** lookupPath_c
             for (size_t i = 0; lookupPath_c[i] != nullptr; i++)
                 lookupPath.push_back(lookupPath_c[i]);
 
-        return new EvalState{nix::EvalState(nix::LookupPath::parse(lookupPath), store->ptr)};
+        void * p = ::operator new(
+            sizeof(EvalState),
+            static_cast<std::align_val_t>(alignof(EvalState)));
```
Comment (roberth) [inline src/libexpr-c/nix_api_expr.cc:112]: This is causing problems for 24.05.
Might a simpler solution be feasible?

https://github.com/NixOS/nixpkgs/pull/324408/checks?check_run_id=27026504754

**Pull request merged**